### PR TITLE
fix(arborist): crash on conflict #5151

### LINKED ISSUE #5151: [BUG] Crash on conflict
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

```
npm ERR! Cannot read properties of null (reading 'edgesOut')

npm ERR! A complete log of this run can be found in:
npm ERR!     .npm/_logs/2022-07-12T14_15_56_150Z-debug-0.log
```

```
87 verbose stack TypeError: Cannot read properties of null (reading 'edgesOut')
87 verbose stack     at [loadPeerSet] (/usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:1359:38)
87 verbose stack     at async [buildDepStep] (/usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:976:11)
87 verbose stack     at async Arborist.buildIdealTree (/usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/build-ideal-tree.js:218:7)
87 verbose stack     at async Promise.all (index 1)
87 verbose stack     at async Arborist.reify (/usr/lib/node_modules/npm/node_modules/@npmcli/arborist/lib/arborist/reify.js:153:5)
87 verbose stack     at async Install.exec (/usr/lib/node_modules/npm/lib/commands/install.js:145:5)
87 verbose stack     at async module.exports (/usr/lib/node_modules/npm/lib/cli.js:78:5)
88 verbose cwd /tmp/crash
89 verbose Linux 5.15.0-40-generic
90 verbose node v18.4.0
91 verbose npm  v8.13.2
92 error Cannot read properties of null (reading 'edgesOut')
```

### Expected Behavior

Expected to see errors about conflicting versions.

### Steps To Reproduce

Steps to reproduce:
```bash
mkdir crash
cd crash
npm install @nestjs/platform-express@8
npm install @nestjs/core
```

### Environment

- npm: 8.13.2
- Node.js: v18.4.0
- OS Name: Ubuntu 22.04 LTS

### PR DESCRIPTION
**Pull request by siemhesda**
An update to fix crash on conflict issue

## References
Closes #5151

### CODE REVIEW

File: workspaces/arborist/lib/arborist/build-ideal-tree.js [hunk 1]
```diff
@@ -1299,7 +1299,7 @@ This is a one-time fix-up, please be patient...
         continue
       }
 
-      const parentEdge = node.parent.edgesOut.get(edge.name)
+      const parentEdge = node.parent.edgesOut.get(edge.name) || node.parent?.edgeOut?.get(edge.name)
```
Comment (wraithgar) [inline workspaces/arborist/lib/arborist/build-ideal-tree.js]: I don't see how this will prevent the exception, let alone fix the bug.

Comment (thecodeinfluencer) [reply workspaces/arborist/lib/arborist/build-ideal-tree.js]: > I don't see how this will prevent the exception, let alone fix the bug.

Updated ...

**Comment by siemhesda** [comment]
The update clears the bug and shows error about the conflicting versions


Before

```
npm ERR! Cannot read properties of null (reading 'edgesOut')

npm ERR! A complete log of this run can be found in:
npm ERR!     .npm/_logs/2023-...-debug-0.log
```


After

```
npm WARN ERESOLVE overriding peer dependency
npm WARN Found: @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN   peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN   1 more (the root project)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN node_modules/@nestjs/platform-express
npm WARN   peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm WARN   node_modules/@nestjs/core
npm WARN   1 more (the root project)
npm WARN ERESOLVE overriding peer dependency
npm WARN Found: @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN   peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN   1 more (the root project)
npm WARN 
npm WARN Could not resolve dependency:
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN node_modules/@nestjs/platform-express
npm WARN   peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm WARN   node_modules/@nestjs/core
npm WARN   1 more (the root project)
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: @nestjs/core@10.2.6
npm ERR! Found: @nestjs/common@8.4.7
npm ERR! node_modules/@nestjs/common
npm ERR!   peer @nestjs/common@"^8.0.0" from @nestjs/platform-express@8.4.7
npm ERR!   node_modules/@nestjs/platform-express
npm ERR!     peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm ERR!     node_modules/@nestjs/core
npm ERR!       peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm ERR!       1 more (the root project)
npm ERR!     @nestjs/platform-express@"^8.4.7" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @nestjs/common@"^10.0.0" from @nestjs/core@10.2.6
npm ERR! node_modules/@nestjs/core
npm ERR!   @nestjs/core@"*" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: @nestjs/common@10.2.6
npm ERR! node_modules/@nestjs/common
npm ERR!   peer @nestjs/common@"^10.0.0" from @nestjs/core@10.2.6
npm ERR!   node_modules/@nestjs/core
npm ERR!     @nestjs/core@"*" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! 
npm ERR! For a full report see:
npm ERR! C:\Users\mbenj\AppData\Local\npm-cache\_logs\2023-09-29T18_48_45_902Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in: C:\Users\mbenj\AppData\Local\npm-cache\_logs\2023-09-29T18_48_45_902Z-debug-0.log
```

File: workspaces/arborist/lib/arborist/build-ideal-tree.js [hunk 2]
```diff
@@ -1299,8 +1299,8 @@ This is a one-time fix-up, please be patient...
         continue
       }
 
-      const parentEdge = node.parent.edgesOut.get(edge.name)
-      const { isProjectRoot, isWorkspace } = node.parent.sourceReference
+      const parentEdge = node.parent?.edgesOut.get(edge.name) 
+      const { isProjectRoot, isWorkspace } = node.parent?.sourceReference || {}
```
Comment (wraithgar) [inline workspaces/arborist/lib/arborist/build-ideal-tree.js]: This line is missing coverage now.

Comment (siemhesda) [reply workspaces/arborist/lib/arborist/build-ideal-tree.js]: Hi @wraithgar 
Could you explain what you mean?

Comment (wraithgar) [reply workspaces/arborist/lib/arborist/build-ideal-tree.js]: Our test coverage for this repo is at 100%, you added a branching path with the `||` that isn't covered by tests.

Comment (siemhesda) [reply workspaces/arborist/lib/arborist/build-ideal-tree.js]: What can be done about it? Because removing the ` || {}` will be unsafe optional chaining.

Comment (wraithgar) [reply workspaces/arborist/lib/arborist/build-ideal-tree.js]: Add a test that covers this scenario. This is the root of the bug, we need to isolate exactly what gets the code in this situation so we can test that we've fixed it.

File: workspaces/arborist/tap-snapshots/test/arborist/rebuild.js.test.cjs [hunk 3]
```diff
@@ -26,7 +26,11 @@ Array [
       "version": "1.0.0",
     },
     "signal": null,
-    "stderr": "stderr",
+    "stderr": String(
+      Debugger attached.
```
Comment (wraithgar) [inline workspaces/arborist/tap-snapshots/test/arborist/rebuild.js.test.cjs]: Not sure why this changed, it does not look right.

File: workspaces/arborist/tap-snapshots/test/shrinkwrap.js.test.cjs [hunk 4]
```diff
@@ -2652,30 +2652,6 @@ Object {
     "node_modules/foo/node_modules/glob": Object {
       "version": "4.0.5",
     },
-    "node_modules/foo/node_modules/glob/node_modules/graceful-fs": Object {
```
Comment (wraithgar) [inline workspaces/arborist/tap-snapshots/test/shrinkwrap.js.test.cjs]: These probably shouldn't get removed here.

**Comment by siemhesda** [comment]
### Step by step breakdown on reproducing and fixing issue #5151

### 1. Initial State

`workspaces/arborist/lib/arborist/build-ideal-tree.js` `1302`

```javascript
const parentEdge = node.parent.edgesOut.get(edge.name);
const { isProjectRoot, isWorkspace } = node.parent.sourceReference;
const isMine = isProjectRoot || isWorkspace;
const conflictOK = this[_force] || (!isMine && !this.#strictPeerDeps);
```

`workspaces/arborist/lib/place-dep.js` `299`

```
  const peer = virtualRoot.children.get(peerEdge.name)
```

Run:

```bash
node . install @nestjs/platform-express@8 --prefix /path/to/folder
node . install @nestjs/core --prefix /path/to/folder
```

Log:

```
npm ERR! Cannot read properties of null (reading 'edgesOut')

npm ERR! A complete log of this run can be found in: \AppData\Local\npm-cache\_logs\2023-10-04T18_00_44_866Z-debug-0.log
```

### 2. First step

`workspaces/arborist/lib/arborist/build-ideal-tree.js` `1302`

```javascript
const parentEdge = node.parent?.edgesOut.get(edge.name);
const { isProjectRoot, isWorkspace } = node.parent.sourceReference;
const isMine = isProjectRoot || isWorkspace;
const conflictOK = this[_force] || (!isMine && !this.#strictPeerDeps);
```

`workspaces/arborist/lib/place-dep.js` `299`

```
  const peer = virtualRoot.children.get(peerEdge.name)
```

Run:

```bash
node . install @nestjs/platform-express@8 --prefix /path/to/folder
node . install @nestjs/core --prefix /path/to/folder
```

Log:

```
npm ERR! Cannot read properties of null (reading 'sourceReference')

npm ERR! A complete log of this run can be found in: \AppData\Local\npm-cache\_logs\2023-10-04T18_07_33_365Z-debug-0.log
```

### 3. Second step

`workspaces/arborist/lib/arborist/build-ideal-tree.js` `1303`

```javascript
const parentEdge = node.parent?.edgesOut.get(edge.name);
const { isProjectRoot, isWorkspace } = node.parent?.sourceReference || {};
const isMine = isProjectRoot || isWorkspace;
const conflictOK = this[_force] || (!isMine && !this.#strictPeerDeps);
```

`workspaces/arborist/lib/place-dep.js` `299`

```
  const peer = virtualRoot.children.get(peerEdge.name)
```

Run:

```bash
node . install @nestjs/platform-express@8 --prefix /path/to/folder
node . install @nestjs/core --prefix /path/to/folder
```

Log:

```
npm WARN ERESOLVE overriding peer dependency
npm WARN Found: @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN   peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN   1 more (the root project)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN node_modules/@nestjs/platform-express
npm WARN   peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm WARN   node_modules/@nestjs/core
npm WARN   1 more (the root project)
npm WARN ERESOLVE overriding peer dependency
npm WARN Found: @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN   peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN   1 more (the root project)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN node_modules/@nestjs/platform-express
npm WARN   peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm WARN   node_modules/@nestjs/core
npm WARN   1 more (the root project)
npm ERR! Cannot read properties of null (reading 'children')

npm ERR! A complete log of this run can be found in: \AppData\Local\npm-cache\_logs\2023-10-04T18_13_02_386Z-debug-0.log
```

### 4. Third step

`workspaces/arborist/lib/arborist/build-ideal-tree.js` `1303`

```javascript
const parentEdge = node.parent?.edgesOut.get(edge.name);
const { isProjectRoot, isWorkspace } = node.parent?.sourceReference || {};
const isMine = isProjectRoot || isWorkspace;
const conflictOK = this[_force] || (!isMine && !this.#strictPeerDeps);
```

`workspaces/arborist/lib/place-dep.js` `299`

```
  const peer = virtualRoot?.children.get(peerEdge.name)
```

Run:

```bash
node . install @nestjs/platform-express@8 --prefix /path/to/folder
node . install @nestjs/core --prefix /path/to/folder
```
Log:
```
npm WARN ERESOLVE overriding peer dependency
npm WARN Found: @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN 1 more (the root project)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN node_modules/@nestjs/platform-express
npm WARN peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN 1 more (the root project)
npm WARN ERESOLVE overriding peer dependency
npm WARN Found: @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN 1 more (the root project)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN node_modules/@nestjs/platform-express
npm WARN peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN 1 more (the root project)
npm ERR! Cannot read properties of null (reading 'children')

npm ERR! A complete log of this run can be found in: \AppData\Local\npm-cache_logs\2023-10-04T18_13_02_386Z-debug-0.log
PS \Documents\gigs\npm> node . install @nestjs/core --prefix /path/to/folder --prefix crash
npm WARN ERESOLVE overriding peer dependency
npm WARN Found: @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN 1 more (the root project)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN node_modules/@nestjs/platform-express
npm WARN peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN 1 more (the root project)
npm WARN ERESOLVE overriding peer dependency
npm WARN Found: @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN 1 more (the root project)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm WARN node_modules/@nestjs/platform-express
npm WARN peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm WARN node_modules/@nestjs/core
npm WARN 1 more (the root project)
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR!
npm ERR! While resolving: @nestjs/core@10.2.6
npm ERR! Found: @nestjs/common@8.4.7
npm ERR! node_modules/@nestjs/common
npm ERR! peer @nestjs/common@"^8.0.0" from @nestjs/platform-express@8.4.7
npm ERR! node_modules/@nestjs/platform-express
npm ERR! peerOptional @nestjs/platform-express@"^8.0.0" from @nestjs/core@8.4.7
npm ERR! node_modules/@nestjs/core
npm ERR! peer @nestjs/core@"^8.0.0" from @nestjs/platform-express@8.4.7
npm ERR! 1 more (the root project)
npm ERR! @nestjs/platform-express@"^8.4.7" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer @nestjs/common@"^10.0.0" from @nestjs/core@10.2.6
npm ERR! node_modules/@nestjs/core
npm ERR! @nestjs/core@"_" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: @nestjs/common@10.2.6
npm ERR! node_modules/@nestjs/common
npm ERR! peer @nestjs/common@"^10.0.0" from @nestjs/core@10.2.6
npm ERR! node_modules/@nestjs/core
npm ERR! @nestjs/core@"_" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR!
npm ERR! For a full report see:
npm ERR! \AppData\Local\npm-cache_logs\2023-10-04T18_21_10_887Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in: \AppData\Local\npm-cache_logs\2023-10-04T18_21_10_887Z-debug-0.log

```

**Comment by siemhesda** [comment]
Hi @wraithgar,
Here is an update on the root cause analysis of the related issue:

### 1. Original Issue

```
mkdir crash
cd crash
npm install @nestjs/platform-express@8
npm install @nestjs/core
```
throws `npm ERR! Cannot read properties of null (reading 'edgesOut')`

### 2. Reverse the installation steps

```
mkdir crash
cd crash
npm install @nestjs/core
npm install @nestjs/platform-express@8
```
correctly shows npm errors about conflicting versions.

### observation

When installing a package that does not have peerDependencies (like `@nestjs/core` in scenario 1) and there are conflicting versions, the `'edgesOut'` error is thrown

### why?

`node` is missing a `.parent` in scenario 1 because `@nestjs/core` does not have peerDependencies as shown from a `console.log(node)` for both scenarios where `isProjecRoot: true` is the difference for scenario 1:

1.

```
{
   "name": "@nestjs/core",
   "version": "10.2.7",
   "location": "node_modules/@nestjs/core",
   "path": "C:\\Users\\mbenj\\Documents\\gigs\\npm\\crash\\node_modules\\@nestjs\\core",
   "resolved": "https://registry.npmjs.org/@nestjs/core/-/core-10.2.7.tgz",
   "extraneous": true,
   "dev": true,
   "optional": true,
   "peer": true,
   "edgesOut": {},
   "edgesIn": {}
}
```

2.

```
{
   "name": "@nestjs/core",
   "version": "10.2.7",
   "location": "",
   "path": "\\npm\\crash\\node_modules\\@nestjs\\core",
   "resolved": "https://registry.npmjs.org/@nestjs/core/-/core-10.2.7.tgz",
   "extraneous": true,
   "dev": true,
   "optional": true,
   "peer": true,
   "isProjectRoot": true,
   "edgesOut": {}
}
```

**Pull request closed without merging**